### PR TITLE
Safe interface for polygons

### PR DESCRIPTION
**Pull request by Lemmih**
Interface changes:
 * `fromPoints` has new constraints: `(Eq r, Num r)`
 * `size :: Polygon t p r -> Int` is new.
 * `unsafeFromPoints :: [Point 2 r :+ p] -> SimplePolygon p r` is new.
 * `unsafeOuterBoundaryVector` is new.
 * `unsafeOuterVertex   :: Int -> Lens' (Polygon t p r) (Point 2 r :+ p)` is new.
 * `unsafeFromCircularVector :: CircularVector (Point 2 r :+ p) -> SimplePolygon p r` is new.
 * `MultiPolygon  :: Polygon Simple p r -> [Polygon Simple p r] -> Polygon Multi  p r` now takes `SimplePolygon` for the outer boundary instead of a circular sequence.
 * `asSimplePolygon` is now `outerBoundary`.
 * `maximumBy :: (Point 2 r :+ p -> Point 2 r :+ p -> Ordering) -> Polygon t p r -> Point 2 r :+ p`
 * `minimumBy :: (Point 2 r :+ p -> Point 2 r :+ p -> Ordering) -> Polygon t p r -> Point 2 r :+ p`
 * `findRotateTo :: (Point 2 r :+ p -> Bool) -> SimplePolygon p r -> Maybe (SimplePolygon p r)`
 * `rotateLeft :: Int -> SimplePolygon p r -> SimplePolygon p r`
 * `rotateRight :: Int -> SimplePolygon p r -> SimplePolygon p r`

All unsafe functions are marked. All safe functions do basic sanity checking: must have at least three points and be in CCW order.

### CODE REVIEW

File: hgeometry/src/Data/Geometry/Polygon/Core.hs [hunk 1]
```diff
@@ -657,3 +680,44 @@ asSimplePolygon (MultiPolygon vs _)  = vs
...
+maximum = F.maximumBy (comparing _core) . view outerBoundaryVector
+
+-- | /O(n)/ Yield the maximum point of a polygon according to the given comparison function.
+maximumBy :: (Point 2 r :+ p -> Point 2 r :+ p -> Ordering) -> SimplePolygon p r -> Point 2 r :+ p
+maximumBy fn = F.maximumBy fn . view outerBoundaryVector
+
+-- | /O(n)/ Yield the maximum point of the polygon.
+minimum :: Ord r => SimplePolygon p r -> Point 2 r :+ p
+minimum = F.minimumBy (comparing _core) . view outerBoundaryVector
+
```
Comment (noinia) [inline hgeometry/src/Data/Geometry/Polygon/Core.hs:700]: these could just have Polygon t p r as type signature no? i.e. hole vertices can't be maximal or minimal anyway.

We should be a bit more precise what the maximum point means though? I mean this seems to produce the lexicographically largest point (max x, max y). I'm not entirely sure why that is so useful to have. Having the 'minimumBY or maximumBy functions seem more useful to me).

Comment (noinia) [reply hgeometry/src/Data/Geometry/Polygon/Core.hs:700]: (note that minimumBy/maxBy do work only for simple polygons; i.e. there we should also iterate over hole vertices) to be correct.

Comment (Lemmih) [reply hgeometry/src/Data/Geometry/Polygon/Core.hs:709]: Good points. I've relaxed the type signatures and written better documentation. Deleting 'minimum' and 'maximum' might be the best option, though.

Comment (Lemmih) [reply hgeometry/src/Data/Geometry/Polygon/Core.hs:706]: I've removed 'minimum' and 'maximum' from the export list.

**Comment by Lemmih** [comment]
I'd like to add a `Traversable` and `Foldable` instance. Think they would complement the `Bifoldable` and `Bitraversable` instance rather than be confusing.

**Comment by Lemmih** [comment]
I can't get pattern synonyms and GADTs working well together. Ideally, I'd like a safe 'SimplePolygon' smart constructor but it doesn't type-check with GADTs. Sad.

**Comment by Lemmih** [comment]
Ok, I figured out how to expose a safe API. I'm just hiding the `CircularVector` behind a newtype and it doesn't get exported. All unchecked modifier functions (eg `unsafeFromPoints`, `unsafeOuterVertex`, `unsafeOuterBoundaryVector`) are now marked as unsafe. All other functions are safe.

I wonder if an extra `Ord r` constraint on the safe functions would be worth it. It would guarantee that the polygon points are unique and no edges are overlapping. What do you think?

**Comment by Lemmih** [comment]
Oh, `Ord r` isn't enough to check for intersections. Hm.

**Comment by noinia** [comment]
> Ok, I figured out how to expose a safe API. I'm just hiding the `CircularVector` behind a newtype and it doesn't get exported. All unchecked modifier functions (eg `unsafeFromPoints`, `unsafeOuterVertex`, `unsafeOuterBoundaryVector`) are now marked as unsafe. All other functions are safe.

Nice!

> I wonder if an extra `Ord r` constraint on the safe functions would be worth it. It would guarantee that the polygon points are unique and no edges are overlapping. What do you think?

> Oh, `Ord r` isn't enough to check for intersections. Hm. Is it possible to write an `O(log n)` algorithm that checks for intersections without finding the actual intersection points? Would it work without a `Fractional r` constraint?

Hmm, just Ord r would be ok I think. But having a Fractional constraint is probably not so nice. I.e. I can certainly imagine wanting to have 'SimplePolygon p Int''s. 

Do you want to test if the polygon boundary intersects itself? I think the best you could hope for there is an O(n) time algorithm. (And in theory O(n) time is possible by trying to construct a triangulation and giving up if you somehow encounter an intersection). That does not seem like a very practical approach though/ I don't think there is an "easy" solution unfortunately.

**Comment by Lemmih** [comment]
> Do you want to test if the polygon boundary intersects itself? I think the best you could hope for there is an O(n) time algorithm. (And in theory O(n) time is possible by trying to construct a triangulation and giving up if you somehow encounter an intersection). That does not seem like a very practical approach though/ I don't think there is an "easy" solution unfortunately.

Ah, sorry, I meant an `O(n log n)` solution.

**Comment by Lemmih** [comment]
Oh oh! I figured out how to check for self-intersections without adding a `Fractional` constraint. Only `Ord` is required. Will update the PR tomorrow.

**Comment by noinia** [comment]
> Oh oh! I figured out how to check for self-intersections without adding a `Fractional` constraint. Only `Ord` is required. Will update the PR tomorrow.

Nice. I guess the Bentley Ottmann sweepline algorithm could be adapted to this end? Testing if the line through points a and b intersects the line through c and d means testing if  a b c and a b d have a different ccw result. I guess if you also do the reverse test (i.e. testing if a and b are on different sides of c d) the line segments intersect.

**Comment by noinia** [comment]
Most of the convex hull algorithms claim that they produce the polygon in CW order. That is probably no longer true now. We should update the docs there. (also now some of the reversing of the vertices before pushing them into a ConvexPolygon there might not be needed anymore).

**Comment by Lemmih** [comment]
Hm, the Foldable instance is weird. Will leave it out.

**Comment by Lemmih** [comment]
Will merge this. As always, let me know if you want anything changed.

**Comment by noinia** [comment]
Hmm, I'm not sure that 'fromPoints' should check for selfIntersections / I think it would be nice to have a version of 'fromPoints' that does make sure the polygon is in CCW and checks vertices are not repeated (since those checks are cheap), but I think that most of the time I would know already that the polygon is not self intersecting (and I would not want to pay the O(n log n) time to check). 

I would suggest to have 'fromPoints' not have the selfIntersection check, thus guaranteeing an O(n) running time, and have a 'safeFromPoints' or so that would also check for selfintersections.

**Pull request merged**